webviz_config/webviz_deprecated.py

```diff
@@ -0,0 +1,174 @@
+from itertools import cycle
+import inspect
+import functools
+import warnings
+
+from decorator import decorator
+
+try:
+    # Python 3.8+
+    from importlib.metadata import distribution  # type: ignore
+except ModuleNotFoundError:
+    # Python < 3.8
+    from importlib_metadata import distribution  # type: ignore
+
+from ._plugin_abc import WebvizPluginABC
+from ._deprecated_argument import DeprecatedArgument
+
+
+def get_class_that_defined_method(
+    method: Any,
+) -> Optional[Any]:
+    if isinstance(method, functools.partial):
+        return get_class_that_defined_method(method.func)
+    if inspect.ismethod(method) or (
+        inspect.isbuiltin(method)
+        and getattr(method, "__self__", None) is not None
+        and getattr(method.__self__, "__class__", None)
+    ):
+        for cls in inspect.getmro(method.__self__.__class__):
+            if method.__name__ in cls.__dict__:
+                return cls
+        method = getattr(method, "__func__", method)  # fallback to __qualname__ parsing
+    if inspect.isfunction(method):
+        cls = getattr(
+            inspect.getmodule(method),
+            method.__qualname__.split(".<locals>", 1)[0].rsplit(".", 1)[0],
+            None,
+        )
+        if isinstance(cls, type):
+            return cls
+    return getattr(method, "__objclass__", None)  # handle special descriptor objects
+
+
+def get_documentation_url(distribution_name: str) -> Optional[str]:
+    dist = distribution(distribution_name)
+    project_urls: Dict[str, str] = {
+        value.split(",")[0]: value.split(",")[1].strip()
+        for (key, value) in dist.metadata.items()
+        if key == "Project-URL"
+    }
+    return project_urls.get("Documentation")
+
+
+def deprecated_plugin(
+    short_message: str = "", long_message: str = ""
+) -> Callable[[WebvizPluginABC], WebvizPluginABC]:
+    def wrapper(original_plugin: WebvizPluginABC) -> WebvizPluginABC:
+        original_plugin.PLUGIN_DEPRECATED = True
+        original_plugin.PLUGIN_DEPRECATION_SHORT_MESSAGE = short_message
+        original_plugin.PLUGIN_DEPRECATION_LONG_MESSAGE = long_message
+
+        members = dict(inspect.getmembers(original_plugin))
+        name = members["__module__"].lower()
+        name_parts = name.split(".")
+        distribution_name = name_parts[0]
+        plugin_name = name_parts[-1].replace("_", "")
+        original_plugin.PLUGIN_DEPRECATION_URL = (
+            f"{get_documentation_url(distribution_name)}"
+            f"/#/deprecations?id={plugin_name}"
+        )
+
+        warnings.warn(
+            """Plugin {} has been deprecated.
+------------------------
+{}
+===
+{}
+""".format(
+                str(original_plugin), short_message, long_message
+            ),
+            DeprecationWarning,
+        )
+
+        return original_plugin
+
+    return wrapper
+
+
+# pylint: disable=dangerous-default-value
+# pylint: disable=keyword-arg-before-vararg
+@decorator
```

Comment:
It keeps the true signature of the function (which is very important for the automatic argument parsing and the new overload decorator): https://stackoverflow.com/questions/13907644/decorator-module-vs-functools-wraps

Reply:
The documentation of `functools.wraps()` states that it does keep the signature of a function since Python 3.4. However, when using it with additional arguments in a shape like the following, it does not appear to keep the original signature (end up with error messages in config_parser):
```python
def decorator(arg):
    def actual_decorator(func):
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            ...do something with arg
            return func(*args, **kwargs)
        return wrapper
    return actual_decorator
```

Reply:
See here for more information: https://hynek.me/articles/decorators/
Unfortunately, `@functools.wraps()` does still not keep the original signature. It does, however, keep the name and the docstring.

Reply:
```python
>>> import functools
>>> def decorator(func):
...     @functools.wraps(func)
...     def wrapper(*args, **kwargs):
...             return func(*args, **kwargs)
...     return wrapper
...
>>> def mult(a, b, c=1):
...     return a * b * c
...
>>> mult_decorated = decorator(mult)
>>> import inspect
>>> inspect.getfullargspec(mult)
FullArgSpec(args=['a', 'b', 'c'], varargs=None, varkw=None, defaults=(1,), kwonlyargs=[], kwonlydefaults=None, annotations={})
>>> inspect.getfullargspec(mult_decorated)
FullArgSpec(args=[], varargs='args', varkw='kwargs', defaults=None, kwonlyargs=[], kwonlydefaults=None, annotations={})
>>> mult.__name__ == mult_decorated.__name__
True
>>> mult.__doc__ == mult_decorated.__doc__
True
>>>```